tests/integration/specs_tests.rs

```diff
@@ -0,0 +1,292 @@
+  /// test directory to a temporary directory before running the
+  /// steps.
+  #[serde(default)]
+  pub temp_dir: bool,
+  /// The base environment to use for the test.
+  #[serde(default)]
+  pub base: Option<String>,
+  /// Whether to use an http server.
+  #[serde(default)]
+  pub http_server: bool,
+  pub steps: Vec<StepMetaData>,
+}
+
+#[derive(Clone, Deserialize)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+struct SingleTestMetaData {
+  #[serde(default)]
+  pub base: Option<String>,
+  #[serde(default)]
+  pub temp_dir: bool,
+  #[serde(default)]
+  pub http_server: bool,
+  #[serde(flatten)]
+  pub step: StepMetaData,
+}
+
+impl SingleTestMetaData {
+  pub fn into_multi(self) -> MultiTestMetaData {
+    MultiTestMetaData {
+      only: self.step.only,
+      base: self.base,
+      temp_dir: self.temp_dir,
+      http_server: self.http_server,
+      steps: vec![self.step],
+    }
+  }
+}
+
+#[derive(Clone, Deserialize)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+struct StepMetaData {
+  #[serde(default)]
+  pub only: bool,
+  /// Whether to clean the deno_dir before running the step.
+  #[serde(default)]
+  pub clean: bool,
+  pub args: VecOrString,
+  pub output: String,
+  #[serde(default)]
+  pub exit_code: i32,
+}
+
+#[derive(Clone)]
+struct Test {
+  pub name: String,
+  pub cwd: PathRef,
+  pub metadata: MultiTestMetaData,
+}
+
+impl Test {
+  pub fn resolve_test_and_assertion_files(&self) -> HashSet<PathRef> {
+    let mut result = HashSet::with_capacity(self.metadata.steps.len() + 1);
+    result.insert(self.cwd.join("__test__.json"));
+    result.extend(
+      self
+        .metadata
+        .steps
+        .iter()
+        .map(|step| self.cwd.join(&step.output)),
+    );
+    result
+  }
+}
+
+struct TestCategory {
+  pub name: String,
+  pub tests: Vec<Test>,
+}
+
+impl TestCategory {
+  pub fn with_only(&self) -> Self {
+    TestCategory {
+      name: self.name.clone(),
+      tests: self
+        .tests
+        .iter()
+        .filter(|test| test.metadata.only)
+        .map(|test| {
+          let mut test = test.clone();
+          if test.metadata.steps.iter().any(|t| t.only) {
+            test.metadata.steps.retain(|t| t.only);
+          }
+          test
+        })
+        .collect(),
+    }
+  }
+}
+
+fn filter(categories: Vec<TestCategory>) -> Vec<TestCategory> {
+  let only_categories = categories
+    .iter()
+    .map(|c| c.with_only())
+    .filter(|c| !c.tests.is_empty())
+    .collect::<Vec<_>>();
+  if !only_categories.is_empty() {
+    if std::env::var("CI").is_ok() {
+      panic!("A test had `\"only\": true` set. Please remove this before committing.");
+    }
+    only_categories
+  } else {
+    categories
+  }
+}
+
+fn collect_tests() -> Vec<TestCategory> {
+  let specs_dir = tests_path().join("specs");
+  let mut result = Vec::new();
+  for entry in specs_dir.read_dir() {
+    let entry = entry.unwrap();
+    if !entry.file_type().unwrap().is_dir() {
```

Comment:
Use `.expect()` instead of unwrap here and provide some useful context. It's gonna be a pain to figure out what's wrong once we hit any of these `unwrap`s